### PR TITLE
MPI Allreduce fallbacks for aarch64 support

### LINKED ISSUE #820: Support ARM / aarch64 / "Apple silicon" M-series processors
I recently got a shiny new toy, and one of the first things I tried was to run some DFTK on it. Unfortunately, it looks like for practical purposes, MPI.jl does not yet work on ARM (see the stacktrace below).

I think the cleanest solution would be to provide a way to completely disable use of MPI (e.g. using `Preferences.jl`). This might be more generally useful, e.g. if MPI is not available on a particular machine.

```
ERROR: User-defined reduction operators are currently not supported on non-Intel architectures.
See https://github.com/JuliaParallel/MPI.jl/issues/404 for more details.
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] MPI.Op(f::Function, T::Type; iscommutative::Bool)
    @ MPI ~/.julia/packages/MPI/APiiL/src/operators.jl:95
  [3] MPI.Op(f::Function, T::Type)
    @ MPI ~/.julia/packages/MPI/APiiL/src/operators.jl:91
  [4] Allreduce!(rbuf::MPI.RBuffer{Base.RefValue{ForwardDiff.Dual{ForwardDiff.Tag{DFTK.var"#409#410"{Float64, DFTK.Smearing.None, PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}, Vector{Vector{Float64}}}, Float64}, Float64, 1}}, Base.RefValue{ForwardDiff.Dual{ForwardDiff.Tag{DFTK.var"#409#410"{Float64, DFTK.Smearing.None, PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}, Vector{Vector{Float64}}}, Float64}, Float64, 1}}}, op::Function, comm::MPI.Comm)
    @ MPI ~/.julia/packages/MPI/APiiL/src/collective.jl:668
  [5] Allreduce!(sendbuf::Base.RefValue{ForwardDiff.Dual{ForwardDiff.Tag{DFTK.var"#409#410"{Float64, DFTK.Smearing.None, PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}, Vector{Vector{Float64}}}, Float64}, Float64, 1}}, recvbuf::Base.RefValue{ForwardDiff.Dual{ForwardDiff.Tag{DFTK.var"#409#410"{Float64, DFTK.Smearing.None, PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}, Vector{Vector{Float64}}}, Float64}, Float64, 1}}, op::Function, comm::MPI.Comm)
    @ MPI ~/.julia/packages/MPI/APiiL/src/collective.jl:670
  [6] Allreduce(obj::ForwardDiff.Dual{ForwardDiff.Tag{DFTK.var"#409#410"{Float64, DFTK.Smearing.None, PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}, Vector{Vector{Float64}}}, Float64}, Float64, 1}, op::Function, comm::MPI.Comm)
    @ MPI ~/.julia/packages/MPI/APiiL/src/collective.jl:695
  [7] mpi_sum(arr::ForwardDiff.Dual{ForwardDiff.Tag{DFTK.var"#409#410"{Float64, DFTK.Smearing.None, PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}, Vector{Vector{Float64}}}, Float64}, Float64, 1}, comm::MPI.Comm)
    @ DFTK ~/.julia/packages/DFTK/Da6eM/src/common/mpi.jl:10
  [8] weighted_ksum(basis::PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}, array::Vector{ForwardDiff.Dual{ForwardDiff.Tag{DFTK.var"#409#410"{Float64, DFTK.Smearing.None, PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}, Vector{Vector{Float64}}}, Float64}, Float64, 1}})
    @ DFTK ~/.julia/packages/DFTK/Da6eM/src/PlaneWaveBasis.jl:484
  [9] excess_n_electrons(basis::PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}, eigenvalues::Vector{Vector{Float64}}, εF::ForwardDiff.Dual{ForwardDiff.Tag{DFTK.var"#409#410"{Float64, DFTK.Smearing.None, PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}, Vector{Vector{Float64}}}, Float64}, Float64, 1}; smearing::DFTK.Smearing.None, temperature::Float64)
    @ DFTK ~/.julia/packages/DFTK/Da6eM/src/occupation.jl:26
 [10] #409
    @ ~/.julia/packages/DFTK/Da6eM/src/occupation.jl:68 [inlined]
 [11] derivative(f::DFTK.var"#409#410"{Float64, DFTK.Smearing.None, PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}, Vector{Vector{Float64}}}, x::Float64)
    @ ForwardDiff ~/.julia/packages/ForwardDiff/QdStj/src/derivative.jl:14
 [12] compute_occupation(basis::PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}, eigenvalues::Vector{Vector{Float64}}, fermialg::FermiTwoStage; tol_n_elec::Float64, temperature::Float64, smearing::DFTK.Smearing.None)
    @ DFTK ~/.julia/packages/DFTK/Da6eM/src/occupation.jl:67
 [13] next_density(ham::Hamiltonian, nbandsalg::AdaptiveBands, fermialg::FermiTwoStage; eigensolver::Function, ψ::Nothing, eigenvalues::Nothing, occupation::Nothing, kwargs::Base.Pairs{Symbol, Real, Tuple{Symbol, Symbol}, NamedTuple{(:miniter, :tol), Tuple{Int64, Float64}}})
    @ DFTK ~/.julia/packages/DFTK/Da6eM/src/scf/self_consistent_field.jl:38
 [14] (::DFTK.var"#fixpoint_map#607"{Float64, DFTK.var"#597#598"{Float64}, χ0Mixing, Float64, typeof(lobpcg_hyper), DFTK.var"#determine_diagtol#601"{Float64}, AdaptiveBands, FermiTwoStage, DFTK.var"#callback#593"{Bool, Bool}, Bool, PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}})(ρin::Array{Float64, 4})
    @ DFTK ~/.julia/packages/DFTK/Da6eM/src/scf/self_consistent_field.jl:126
 [15] (::DFTK.var"#anderson#564"{DFTK.var"#anderson#563#565"{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, Int64}})(f::DFTK.var"#fixpoint_map#607"{Float64, DFTK.var"#597#598"{Float64}, χ0Mixing, Float64, typeof(lobpcg_hyper), DFTK.var"#determine_diagtol#601"{Float64}, AdaptiveBands, FermiTwoStage, DFTK.var"#callback#593"{Bool, Bool}, Bool, PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}}, x0::Array{Float64, 4}, max_iter::Int64; tol::Float64)
    @ DFTK ~/.julia/packages/DFTK/Da6eM/src/scf/scf_solvers.jl:46
 [16] macro expansion
    @ ~/.julia/packages/DFTK/Da6eM/src/scf/self_consistent_field.jl:155 [inlined]
 [17] self_consistent_field(basis::PlaneWaveBasis{Float64, Float64, Array{StaticArraysCore.SVector{3, Int64}, 3}, Array{StaticArraysCore.SVector{3, Float64}, 3}, Vector{StaticArraysCore.SVector{3, Int64}}}; ρ::Array{Float64, 4}, ψ::Nothing, tol::Float64, is_converged::DFTK.var"#597#598"{Float64}, maxiter::Int64, mixing::χ0Mixing, damping::Float64, solver::DFTK.var"#anderson#564"{DFTK.var"#anderson#563#565"{Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, Int64}}, eigensolver::typeof(lobpcg_hyper), determine_diagtol::DFTK.var"#determine_diagtol#601"{Float64}, nbandsalg::AdaptiveBands, fermialg::FermiTwoStage, callback::DFTK.var"#callback#593"{Bool, Bool}, compute_consistent_energies::Bool, response::ResponseOptions)
    @ DFTK ~/.julia/packages/TimerOutputs/LHjFw/src/TimerOutput.jl:237
 [18] top-level scope
    @ REPL[10]:24```

### PR DESCRIPTION
**Pull request by azadoks**
There must be a better place to check and error if there are multiple MPI processes?

Closes #820.

### CODE REVIEW

**Comment by azadoks** [comment]
With some feedback from @mfherbst, I swapped the dispatch for a `@static if` to condition the function definition at compile time on the architecture. It also seems that on one process, the MPI operations should all be no-ops, so I've changed that as well (we're not sure why it was working before?).

It still seems a bit awkward to check that the number of processes is 1 on every call, but because it's a MPI COMM-related property, I'm not sure what would be better.

**Comment by mfherbst** [comment]
> It still seems a bit awkward to check that the number of processes is 1 on every call, but because it's a MPI COMM-related property, I'm not sure what would be better.

I guess we could sneak it in here

https://github.com/azadoks/DFTK.jl/blob/9259c96460eaa27bc0af7ad93e52ae62a7ee33a4/src/PlaneWaveBasis.jl#L223-L243

 via an appropriate function in `common/mpi.jl` that checks MPI is "valid".

**Comment by mfherbst** [comment]
Looks good. Could you add a comment and link to the appropriate issue in MPI.jl ?

File: src/PlaneWaveBasis.jl [hunk 1]
```diff
@@ -220,6 +220,12 @@ function PlaneWaveBasis(model::Model{T}, Ecut::Number, fft_size, variational,
     # by the same process
     n_kpt   = length(kcoords_global)
     n_procs = mpi_nprocs(comm_kpts)
+    # Custom reduction operators for MPI are currently not working on aarch64, so
+    # fallbacks are defined in common/mpi.jl. For them to work, there cannot be more
+    # than 1 MPI process.
+    if Base.Sys.ARCH == :aarch64 && n_procs > 1
+        error("MPI not supported on aarch64")
```
Comment (antoine-levitt) [inline src/PlaneWaveBasis.jl:227]: Move that check to MPI init?

File: src/common/mpi.jl [hunk 2]
```diff
@@ -7,11 +7,25 @@ Number of processors used in MPI. Can be called without ensuring initialization.
...
+    # https://github.com/JuliaParallel/MPI.jl/issues/404), so we define fallback no-op
+    # mpi_* functions to get things working while waiting for an upstream solution.
+    mpi_sum(  arr, ::MPI.Comm) = arr
+    mpi_sum!( arr, ::MPI.Comm) = arr
+    mpi_min(  arr, ::MPI.Comm) = arr
+    mpi_min!( arr, ::MPI.Comm) = arr
+    mpi_max(  arr, ::MPI.Comm) = arr
+    mpi_max!( arr, ::MPI.Comm) = arr
+    mpi_mean( arr, ::MPI.Comm) = arr
+    mpi_mean!(arr, ::MPI.Comm) = arr
```
Comment (antoine-levitt) [inline src/common/mpi.jl:21]: Just FYI (not necessarily to do it here), a nice metaprogramming pattern here is `for fun in (:mpi_sum, :mpi_sum!, ...) @eval $fun(x) = x`

File: src/PlaneWaveBasis.jl (hunk 1 above)
Comment (antoine-levitt) [reply src/PlaneWaveBasis.jl:227]: By that I mean: define our own mpi_init

File: src/common/mpi.jl (hunk 2 above)
Comment (antoine-levitt) [reply src/common/mpi.jl:21]: Also, just stylewise, it's maybe better to put the common case first, and the workaround later, ie `if ARCH != :aarch64`

File: src/PlaneWaveBasis.jl (hunk 1 above)
Comment (azadoks) [reply src/PlaneWaveBasis.jl:227]: `MPI.Init()` doesn't require knowledge of a specific communicator, but _we_ need to have the MPI communicator used by the k-points to check the number of processes it's using. I think that in practice, this is always `COMM_WORLD` (by default), but it doesn't have to be.
Michael had the same suggestion at first, but I'm not sure that it works out.

Comment (antoine-levitt) [reply src/PlaneWaveBasis.jl:227]: For this workaround, we need comm world to have rank 1 so I think it's fine?

**Comment by antoine-levitt** [comment]
Also, link to the issue number in the error message directly, so people can see why MPI is not supported?

File: src/PlaneWaveBasis.jl [hunk 3]
```diff
@@ -220,27 +220,38 @@ function PlaneWaveBasis(model::Model{T}, Ecut::Number, fft_size, variational,
...
+                  "(see https://github.com/JuliaParallel/MPI.jl/issues/404)")
+        end
     end
+
+    # get the slice of 1:n_kpt to be handled by this process
+    # Note: MPI ranks are 0-based
+    krange_allprocs = split_evenly(1:n_kpt, n_procs)
+    krange_thisproc = krange_allprocs[1 + MPI.Comm_rank(comm_kpts)]
+    @assert mpi_sum(length(krange_thisproc), comm_kpts) == n_kpt
+    @assert !isempty(krange_thisproc)
```
Comment (mfherbst) [inline src/PlaneWaveBasis.jl:254]: Be careful this overwrites some special casing we need for the case of your line 230 (CI environment variable)

**Comment by azadoks** [comment]
I think leaving the ARCH check how it is is the easiest solution given that this will (hopefully) be a temporary fix. I can't let `mpi_sum` be called before I do this check, so I'd have to refactor the following logic, which I think in the end would be more confusing than just checking this case first.

Thoughts?

**Comment by mfherbst** [comment]
I agree we should have our own init wrapper, but I think this current state is fine for now.

**Pull request merged**